File: spark/src/main/scala/geotrellis/spark/tiling/FloatingLayoutScheme.scala
package geotrellis.spark.tiling


import geotrellis.raster._
import geotrellis.vector.{Extent, ProjectedExtent}

object FloatingLayoutScheme {
  val DEFAULT_TILE_SIZE = 256

  def apply(): FloatingLayoutScheme =
    apply(DEFAULT_TILE_SIZE)

  def apply(tileSize: Int): FloatingLayoutScheme =
    apply(tileSize, tileSize)

  def apply(tileCols: Int, tileRows: Int): FloatingLayoutScheme =
    new FloatingLayoutScheme(tileCols, tileRows)
}

class FloatingLayoutScheme(tileCols: Int, tileRows: Int) extends LayoutScheme {
  def levelFor(extent: Extent, cellSize: CellSize) =
    0 -> LayoutDefinition(GridExtent(extent, cellSize), tileCols, tileRows)

  // TODO: Fix type system so that there is no runtime error
  def zoomOut(level: LayoutLevel) =
    throw new UnsupportedOperationException("zoomOut not supported for FloatingLayoutScheme")

  def zoomIn(level: LayoutLevel) =
    throw new UnsupportedOperationException("zoomIn not supported for FloatingLayoutScheme")
}

File: geotrellis-python/geotrellis/spark/tiling/FloatingLayoutScheme.py
from __future__ import absolute_import
from geotrellis.spark.tiling.LayoutScheme import LayoutScheme, LayoutLevel
from geotrellis.spark.tiling.LayoutDefinition import LayoutDefinition
from geotrellis.raster.GridExtent import GridExtent

class FloatingLayoutScheme(LayoutScheme):
    DEFAULT_TILE_SIZE = 256
    def __init__(self, tileCols = None, tileRows = None):
        if tileRows is None:
            if tileCols is None:
                tileCols = FloaintLayoutScheme.DEFAULT_TILE_SIZE
                tileRows = FloaintLayoutScheme.DEFAULT_TILE_SIZE
            else:
                tileRows = tileCols
        self.tileCols = tileCols
        self.tileRows = tileRows

    def levelFor(self, extent, cellSize):
        return LayoutLevel(
                0,
                LayoutDefinition(
                    GridExtent(extent, cellSize),
                    self.tileCols,
                    self.tileRows))

    def zoomOut(level):
        raise Exception("zoomOut not supported for FloatingLayoutScheme")
    def zoomIn(level):
        raise Exception("zoomIn not supported for FloatingLayoutScheme")
